File: FutureVoice/Views/ReferralJoinSheet.swift
import SwiftUI

/// "Your friend joined." Shown to the INVITER when someone signs up with
/// their code — the grant lands in a balance nobody watches, so without this
/// the whole loop is invisible from the giving side.
///
/// Presented from `ReferralInbox` by `RootTabView`, after the foreground poll
/// in `ReferralService.announceJoins()` (there is no push infrastructure, so
/// the local notification and this sheet are the whole delivery).
struct ReferralJoinSheet: View {
    @Environment(\.dismiss) private var dismiss
    let join: ReferralJoin
    @State private var revealed = false

    var body: some View {
        VStack(spacing: 28) {
            Spacer()

            Image(systemName: "person.2.fill")
                .font(.system(size: 44, weight: .semibold))
                .foregroundStyle(.tint)
                .opacity(revealed ? 1 : 0)

            VStack(spacing: 12) {
                Text(headline)
                    .font(.title2.weight(.semibold))
                    .multilineTextAlignment(.center)

                if join.minutesEarned > 0 {
                    Text(explain("+\(join.minutesEarned) min"))
                        .font(.largeTitle.weight(.bold))
                        .monospacedDigit()
                        .opacity(revealed ? 1 : 0)
                        .scaleEffect(revealed ? 1 : 0.85)
                }

                Text(support)
                    .font(.subheadline)
                    .foregroundStyle(.secondary)
                    .multilineTextAlignment(.center)
                    .padding(.horizontal, 24)
            }

            Spacer()

            Button {
                dismiss()
            } label: {
                Text("Nice")
                    .frame(maxWidth: .infinity)
            }
            .buttonStyle(.borderedProminent)
            .controlSize(.large)
            .padding(.horizontal, 20)
            .padding(.bottom, 12)
        }
        .onAppear {
            withAnimation(.easeOut(duration: 0.5).delay(0.15)) { revealed = true }
        }
        .presentationDetents([.medium])
    }

    private var headline: String {
        if let name = join.friendName {
            return explain("\(name) joined with your code")
        }
        return join.friendsJoined == 1
            ? explain("A friend joined with your code")
            : explain("\(join.friendsJoined) friends joined with your code")
    }

    /// Past the cap the friend still got their half, so the line stays warm
    /// instead of reading as a rejection.
    private var support: String {
        join.minutesEarned > 0
            ? explain("They got \(ReferralService.bonusMinutes) minutes too. That's \(join.totalJoined) of your \(ReferralService.rewardedInviteCap) rewarded invites.")
            : explain("They got \(ReferralService.bonusMinutes) minutes. Your \(ReferralService.rewardedInviteCap) rewarded invites are already used.")
    }
}

File: android/app/src/main/java/com/roro/futurevoice/ui/ReferralJoinSheet.kt
package com.roro.futurevoice.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.roro.futurevoice.R
import com.roro.futurevoice.data.ReferralJoins
import com.roro.futurevoice.net.ReferralClient

/** A friend joined with your code — said in the app, not only in a notification. */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ReferralJoinSheet(join: ReferralJoins.Join, onDismiss: () -> Unit) {
    val headline = when {
        join.friendName != null -> stringResource(R.string.referral_joined_named, join.friendName)
        join.friendsJoined == 1 -> stringResource(R.string.a_friend_joined_with_your_code_975008)
        else -> stringResource(R.string.lld_friends_joined_with_your_code_fe4dfc, join.friendsJoined)
    }
    // Past the cap the friend still got their half, so the line stays warm
    // instead of reading as a rejection.
    val support = if (join.minutesEarned > 0)
        stringResource(R.string.they_got_lld_minutes_too_that_s_lld_of_your_lld_rewarded_inv_9c811c, ReferralClient.bonusMinutes, join.totalJoined, ReferralClient.REWARDED_INVITE_CAP)
    else stringResource(R.string.they_got_lld_minutes_your_lld_rewarded_invites_are_already_u_5cb510, ReferralClient.bonusMinutes, ReferralClient.REWARDED_INVITE_CAP)
    ModalBottomSheet(onDismissRequest = onDismiss) {
        Column(Modifier.fillMaxWidth().navigationBarsPadding().padding(horizontal = 24.dp).padding(bottom = 32.dp, top = 8.dp),
            horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(14.dp)) {
            Text(headline, style = MaterialTheme.typography.titleLarge, fontWeight = FontWeight.SemiBold)
            if (join.minutesEarned > 0) Text(stringResource(R.string.lld_min_7c4bb6, join.minutesEarned),
                style = MaterialTheme.typography.headlineMedium, fontWeight = FontWeight.Bold,
                color = MaterialTheme.colorScheme.primary)
            Text(support, style = MaterialTheme.typography.bodyMedium, color = MaterialTheme.colorScheme.onSurfaceVariant)
            Button(onClick = onDismiss, modifier = Modifier.fillMaxWidth()) { Text(stringResource(R.string.nice)) }
        }
    }
}
